src/jsdoc_transformer.ts

```diff
@@ -0,0 +1,937 @@
+          default:
+            break;
+        }
+        return ts.visitEachChild(node, visitor, context);
+      }
+
+      return visitor(sourceFile) as ts.SourceFile;
+    };
+  };
+}
+
+/**
+ * jsdocTransformer returns a transformer factory that converts TypeScript types into the equivalent
+ * JSDoc annotations.
+ */
+export function jsdocTransformer(
+    host: AnnotatorHost, tsOptions: ts.CompilerOptions, tsHost: ts.CompilerHost,
+    typeChecker: ts.TypeChecker, diagnostics: ts.Diagnostic[]):
+    (context: ts.TransformationContext) => ts.Transformer<ts.SourceFile> {
+  return (context: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
+    return (sourceFile: ts.SourceFile) => {
+      const moduleTypeTranslator = new ModuleTypeTranslator(
+          sourceFile, typeChecker, host, diagnostics, /*isForExterns*/ false);
+      /**
+       * The set of all names exported from an export * in the current module. Used to prevent
+       * emitting duplicated exports. The first export * takes precedence in ES6.
+       */
+      const expandedStarImports = new Set<string>();
+
+      function visitClassDeclaration(classDecl: ts.ClassDeclaration): ts.Statement[] {
+        const mjsdoc = moduleTypeTranslator.getMutableJSDoc(classDecl);
+        if (transformerUtil.hasModifierFlag(classDecl, ts.ModifierFlags.Abstract)) {
+          mjsdoc.tags.push({tagName: 'abstract'});
+        }
+
+        maybeAddTemplateClause(mjsdoc.tags, classDecl);
+        if (!host.untyped) {
+          maybeAddHeritageClauses(mjsdoc.tags, moduleTypeTranslator, classDecl);
+        }
+        mjsdoc.updateComment();
+        const decls: ts.Statement[] = [];
+        const memberDecl = createMemberTypeDeclaration(moduleTypeTranslator, classDecl);
+        // WARNING: order is significant; we must create the member decl before transforming away
+        // parameter property comments when visiting the constructor.
+        decls.push(ts.visitEachChild(classDecl, visitor, context));
+        if (memberDecl) decls.push(memberDecl);
+        return decls;
+      }
+
+      /**
+       * visitHeritageClause works around a Closure Compiler issue, where the expression in an
+       * "extends" clause must be a simple identifier, and in particular must not be a parenthesized
+       * expression.
+       *
+       * This is triggered when TS code writes "class X extends (Foo as Bar) { ... }", commonly done
+       * to support mixins. For extends clauses in classes, the code below drops the cast and any
+       * parentheticals, leaving just the original expression.
+       *
+       * This is an incomplete workaround, as Closure will still bail on other super expressions,
+       * but retains compatibility with the previous emit that (accidentally) dropped the cast
+       * expression.
+       *
+       * TODO(martinprobst): remove this once the Closure side issue has been resolved.
```

Comment:
Is there a bug on this?

Reply:
There's an internal one (b/111487374); I'm never quite sure where to file bugs with Closure.